### PR TITLE
qnnpack hardswish - LUTs

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36545 hardswish: make it work in static quantization
* #36320 qnnpack hardswish - pytorch op integration
* **#36252 qnnpack hardswish - LUTs**

Summary:

Adds a baseline hardswish kernel using LUTs in QNNPACK.
Performance is 1.9 GB/s on a Nexus 6 and 2.2 GB/s on Pixel 3 - same as other LUT based ops.

Enforcing scale and zp to be equal to the input, to match the server implementation.

There are some potential  improvements in rewriting this as NEON
kernels for a further speedup - saving that until later, if we need it.

Test Plan:

```
with-proxy ./scripts/build-local.sh
./build/local/hardswish-test

with-proxy scripts/build-android-armv7.sh
adb push ./build/android/armeabi-v7a/hardswish-* /data/qnnpack
adb shell
/data/qnnpack/hardswish-test
/data/qnnpack/hardswish-bench

with-proxy scripts/build-android-arm64.sh
adb push ./build/android/arm64-v8a/hardswish-* /data/qnnpack
/data/qnnpack/hardswish-test
/data/qnnpack/hardswish-bench
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20965044](https://our.internmc.facebook.com/intern/diff/D20965044)

### CODE REVIEW

**Comment by vkuzo** [comment]
note: @kimishpatel , @AshkanAliabadi  and myself aligned offline that LUTs are a good approach for now (as opposed to a custom kernel)

**Comment by AshkanAliabadi** [comment]
LGTM.  I think more const-correctness would be great but that's just my personal preference and merely a suggestion. Thanks.

**Comment by supriyar** [comment]
> note: @kimishpatel , @AshkanAliabadi and myself aligned offline that LUTs are a good approach for now (as opposed to a custom kernel)

IIRC this will cause a perf hit when used with pytorch as we will be creating the LUT for each operator invocation. Are we okay with that perf hit for now and do we have plans on optimizing the LUT creation?

**Comment by vkuzo** [comment]
> > note: @kimishpatel , @AshkanAliabadi and myself aligned offline that LUTs are a good approach for now (as opposed to a custom kernel)
> 
> IIRC this will cause a perf hit when used with pytorch as we will be creating the LUT for each operator invocation. Are we okay with that perf hit for now and do we have plans on caching the LUT creation?

100% agree.  Please feel free to check out https://fb.quip.com/EzpbA3BHXxKY for the discussions so far.  The current working hypothesis is that LUTs will deliver a speedup compared to dq -> float -> q for reasonable layer sizes, even with the init every time.  We can wait on reviewing this until we verify with on device performance data (which should be in the next day or two)

File: aten/src/ATen/native/quantized/cpu/qnnpack/bench/hardswish.cc [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  pytorch_qnnp_operator_t hardswishOperator = nullptr;
+  status = pytorch_qnnp_create_hardswish_nc_q8(
+      channels,
+      127 /* input zero point */,
+      1.0f /* input scale */,
+      127 /* output zero point */,
+      1.0f /* output scale */,
+      0 /* output min */,
```
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/qnnpack/bench/hardswish.cc:44]: Can we use `std::numeric_limits` for min and max calculations here?

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/hardswish.c [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+  }
+
+  hardswish_op->lookup_table = malloc(256 * sizeof(uint8_t));
+  if (hardswish_op->lookup_table == NULL) {
+    pytorch_qnnp_log_error(
+        "failed to allocate 256 bytes for Hardswish lookup table");
+    goto error;
+  }
+
+  uint8_t* lookup_table = hardswish_op->lookup_table;
```
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/qnnpack/src/hardswish.c:104]: I wonder if there is a requirement or interest in supporting other data types for the LUT. But that would involve re-writing the kernel, I think. Might be good to mark it as a TODO (if server supports it)

Comment (vkuzo) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/hardswish.c:104]: from what I understand uint8_t is the only dtype supported in QNNPACK as a whole

**Review by supriyar — APPROVED** [review]
Looks Good! Please add the on-device benchmark numbers to the PR once you have them.

**Comment by vkuzo** [comment]
> Looks Good! Please add the on-device benchmark numbers to the PR once you have them.

I'll comment with them on the Facebook diff, since they depend on internal infra.

File: aten/src/ATen/native/quantized/cpu/qnnpack/src/hardswish.c (hunk 2 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qnnpack/src/hardswish.c:104]: @vkuzo, btw another thought about avoiding to build lookup table: You can have a model level parameter or attribute, that stores all the lookup tables that caches on quantization parameters. So now you have `quantized::lookup_lookup_table` op that given the parameters returns lookup table either wrapped in a Tensor (via cpp_type_hack) or via custom class (via torchbind). This op takes in lookup_table as input and check if the lookup table exist, else it builds one. Not too different from current conv_prepack.

**Pull request closed without merging**